src/AutoMapper/Mappers/CollectionMapper.cs

```diff
@@ -48,11 +52,84 @@ public bool IsMatch(TypePair context)
             return isMatch;
         }
 
-        public Expression MapExpression(TypeMapRegistry typeMapRegistry, IConfigurationProvider configurationProvider, Expression sourceExpression, Expression destExpression, Expression contextExpression)
+        public Expression MapExpression(TypeMapRegistry typeMapRegistry, IConfigurationProvider configurationProvider, PropertyMap propertyMap, Expression sourceExpression, Expression destExpression, Expression contextExpression)
         {
-            return Expression.Call(null, 
-                MapMethodInfo.MakeGenericMethod(sourceExpression.Type, TypeHelper.GetElementType(sourceExpression.Type), destExpression.Type, TypeHelper.GetElementType(destExpression.Type)),
-                    sourceExpression, destExpression, contextExpression);
+
+            var sourceElementType = TypeHelper.GetElementType(sourceExpression.Type);
+            var destElementType = TypeHelper.GetElementType(destExpression.Type);
+
+            var typePair = new TypePair(sourceElementType, destElementType);
+            var typeMap = typeMapRegistry.GetTypeMap(typePair);
+
+
+            var destCollectionType = destExpression.Type;
+
+            var makeGenericType = typeof(List<>).MakeGenericType(destElementType);
+            var destICollectionType = typeof(ICollection<>).MakeGenericType(destElementType);
+            var newExpr = ToType(
+                destExpression.Type.IsInterface()
+                    ? New(makeGenericType)
+                    : DelegateFactory.GenerateConstructorExpression(destExpression.Type),
+                destICollectionType);
+            var ifNullExpr = configurationProvider.AllowNullCollections
+                     ? Constant(null, destCollectionType)
+                     : newExpr;
+
+            var itemParam = Parameter(sourceElementType, "item");
+            var itemContextParam = Parameter(typeof(ResolutionContext), "itemContext");
+
+            var blockExprs = new List<Expression>();
+            var blockParams = new List<ParameterExpression>();
+            if (propertyMap.UseDestinationValue)
+            {
+                blockExprs.Add(Call(destExpression, destICollectionType.GetMethod("Clear")));
+            }
+            else
+            {
+                var destParam = Parameter(newExpr.Type, "dest");
+                blockParams.Add(destParam);
+                blockExprs.Add(Assign(destParam, newExpr));
+                destExpression = destParam;
+            }
+
+            Expression itemExpr;
+
+            var match = configurationProvider.GetMappers().FirstOrDefault(m => m.IsMatch(typePair));
```

Comment:
This looks like duplication of code.  Maybe make a function in typeMap pairing that has only this code so you don't have two version of this.  Like `GetMapExpression(IEnumerable<IObjectMapper> mappers, TypePair typePair)` or something like that.